capsules/extra/src/sht4x.rs

```diff
@@ -0,0 +1,261 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2023.
+
+//! Driver for SHT4x Temperature and Humidity Sensor
+
+use core::cell::Cell;
+use enum_primitive::cast::FromPrimitive;
+use enum_primitive::enum_from_primitive;
+use kernel::hil::i2c;
+use kernel::hil::time::{self, Alarm, ConvertTicks};
+use kernel::utilities::cells::{OptionalCell, TakeCell};
+use kernel::ErrorCode;
+
+pub static BASE_ADDR: u8 = 0x44;
+
+enum_from_primitive! {
+    enum Registers {
+        /// sht4x has no Clock Stretching
+        /// Measurement High Repeatability
+        MEASHIGHREP = 0xFD,
+        /// Measurement Medium Repeatability
+        MEASMEDREP = 0xF6,
+        /// Measurement Low Repeatability
+        MEASLOWREP = 0xE0,
+        /// Read Serial Number
+        READSERIALNUM = 0x89,
+        /// Soft Reset
+        SOFTRESET = 0x94,
+        /// Activate heater with 200mW for 1s
+        HEATER200MW1S = 0x39,
+        /// Activate heater with 200mW for 0.1s
+        HEATER200MW01S = 0x32,
+        /// Activate heater with 110mW for 1s
+        HEATER110MW1S = 0x2F,
+        /// Activate heater with 110mW for 0.1s
+        HEATER110MW01S = 0x24,
+        /// Activate heater with 20mW for 1s
+        HEATER20MW1S = 0x1E,
+        /// Activate heater with 20mW for 0.1s
+        HEATER20MW01S = 0x15,
+    }
+}
+
+#[derive(Clone, Copy, PartialEq)]
+enum State {
+    Idle,
+    Read,
+    ReadData,
+}
+
+fn crc8(data: &[u8]) -> u8 {
+    let polynomial = 0x31;
+    let mut crc = 0xff;
+
+    for x in 0..data.len() {
+        crc ^= data[x];
+        for _i in 0..8 {
+            if (crc & 0x80) != 0 {
+                crc = crc << 1 ^ polynomial;
+            } else {
+                crc = crc << 1;
+            }
+        }
+    }
+    crc
+}
+
+pub struct SHT4x<'a, A: Alarm<'a>, I: i2c::I2CDevice> {
+    i2c: &'a I,
+    humidity_client: OptionalCell<&'a dyn kernel::hil::sensors::HumidityClient>,
+    temperature_client: OptionalCell<&'a dyn kernel::hil::sensors::TemperatureClient>,
+    state: Cell<State>,
+    buffer: TakeCell<'static, [u8]>,
+    read_temp: Cell<bool>,
+    read_hum: Cell<bool>,
+    alarm: &'a A,
+}
+
+impl<'a, A: Alarm<'a>, I: i2c::I2CDevice> SHT4x<'a, A, I> {
+    pub fn new(i2c: &'a I, buffer: &'static mut [u8], alarm: &'a A) -> SHT4x<'a, A, I> {
+        SHT4x {
+            i2c: i2c,
+            humidity_client: OptionalCell::empty(),
+            temperature_client: OptionalCell::empty(),
+            state: Cell::new(State::Idle),
+            buffer: TakeCell::new(buffer),
+            read_temp: Cell::new(false),
+            read_hum: Cell::new(false),
+            alarm: alarm,
+        }
+    }
+
+    fn read_humidity(&self) -> Result<(), ErrorCode> {
+        if self.read_hum.get() == true {
+            Err(ErrorCode::BUSY)
+        } else {
+            if self.state.get() == State::Idle {
+                self.read_hum.set(true);
+                self.read_temp_hum()
```

Comment:
I don't think we can return the result of this function directly. If an error occurs, I don't think we want `self.read_hum` to be set to true.